### PR TITLE
starlight tool file updated

### PR DESCRIPTION
**Pull request by KiSooLee**
cmssw-tool-conf.spec issue is resolved

### CODE REVIEW

File: starlight.spec [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Release:	1%{?dist}
```
Comment (ghost) [inline starlight.spec]: This definition is unused, it should instead read:
`### RPM external starlight r193` (using the current release on cms-externals repository)

File: starlight.spec [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+%patch0 -p1
+
+%build
+%configure
+make %{?_smp_mflags}
+
+
+%install
+cmake <path-to-starlight-source> -DENABLE_DPMJET=on
```
Comment (ghost) [inline starlight.spec]: This line needs to be completed.

File: starlight.spec [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+
+%prep
+%setup -q -n %{n}/%{realversion}
+
+%patch0 -p1
+
+%build
+%configure
+make %{?_smp_mflags}
```
Comment (ghost) [inline starlight.spec]: It doesn't seem that this variable is defined here.

File: starlight.spec [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+%define keep_archives true
+
+%description
+
+
+%prep
+%setup -q -n %{n}/%{realversion}
+
+%patch0 -p1
```
Comment (ghost) [inline starlight.spec]: You are not defining any patch, this will fail.

**Comment by ghost** [comment]
Also please test this following this guide: https://twiki.cern.ch/twiki/bin/view/CMS/SDTHowToBuildExternalTools

**Comment by KiSooLee** [comment]
Thank you for the guide

File: starlight.spec (hunk 2 above)
Comment (KiSooLee) [inline starlight.spec]: how can I complete it?

cmake <path-to-starlight-source> -DENABLE_DPMJET=on
make

with these two line the package is compiled in local

Comment (KiSooLee) [inline starlight.spec]: Oh my mistake.
<path-to-starlight-source> should be modified

**Comment by KiSooLee** [comment]
+1

File: starlight.spec [hunk 5]
```diff
@@ -0,0 +1,19 @@
+### RPM external starlight r193
+Requires: clhep gfortran
```
Comment (davidlt) [inline starlight.spec]: Remove `gfortran`, there are no separate external for it (it comes from GCC package).

File: starlight.spec [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+%define branch cms/%{realversion}
+%define github_user cms-externals
+Source0:	git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
```
Comment (davidlt) [inline starlight.spec]: Remove extra whitespace.

File: starlight-toolfile.spec [hunk 7]
```diff
@@ -0,0 +1,33 @@
+### RPM external starlight-toolfile 1.0
+Requires:	starlight
+
+%description
```
Comment (davidlt) [inline starlight-toolfile.spec]: Remove, this should be filled by cmsBuild.

File: starlight-toolfile.spec [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+
+%prep
+%setup -q
+
+
+%build
```
Comment (davidlt) [inline starlight-toolfile.spec:6]: I think, you also don't need this one. Default will be supplied by cmsBuild.

File: starlight-toolfile.spec [hunk 9]
```diff
@@ -0,0 +1,33 @@
+### RPM external starlight-toolfile 1.0
+Requires:	starlight
+
+%description
+
+
+%prep
+%setup -q
```
Comment (davidlt) [inline starlight-toolfile.spec]: Please, don't silence it. It might help with debugging if something goes wrong.

File: starlight.spec [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+BuildRequires:	cmake
+
+%define keep_archives true
+
+%prep
+%setup -q -n %{n}/%{realversion}
```
Comment (davidlt) [inline starlight.spec]: Same here, don't need to silence `%setup`.

File: starlight-toolfile.spec [hunk 11]
```diff
@@ -0,0 +1,33 @@
+### RPM external starlight-toolfile 1.0
+Requires:	starlight
```
Comment (davidlt) [inline starlight-toolfile.spec]: Remove extra whitespace.

File: starlight-toolfile.spec [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  <client>
+    <environment name="STARLIGHT_BASE" default="@TOOL_ROOT@"/>
+    <environment name="LIBDIR" default="$STARLIGHT_BASE/lib"/>
+    <environment name="INCLUDE" default="$STARLIGHT_BASE/include"/>
+  </client>
+  <runtime name="STARLIGHTDATA" value="$STARLIGHT_BASE/share/Starlight/xmldoc"/>
+  <runtime name="ROOT_INCLUDE_PATH" value="$INCLUDE" type="path"/>
+  <use name="root_cxxdefaults"/>
+  <use name="cxxcompiler"/>
+  <use name="gfortran"/>
```
Comment (davidlt) [inline starlight-toolfile.spec]: You don't need `cxxcompiler` and `gfortran`.

File: starlight.spec [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+%build
+
+%install
+cmake ./
+make
```
Comment (davidlt) [inline starlight.spec]: CMake goes into `%build` and contains 'make %{makeprocesses}'. `%install` should have `make %{make processes} install`.

File: starlight.spec (hunk 10 above)
Comment (KiSooLee) [inline starlight.spec]: Could you please let me know the meaning of silence?
The option -p or %?

Comment (davidlt) [inline starlight.spec]: `-q`

**Comment by ghost** [comment]
@cmsbuild please test

File: starlight.spec [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+Source0:git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
+
+BuildRequires:	cmake
+
+%define keep_archives true
+
+%prep
+%setup -n %{n}/%{realversion}
+
+%build
```
Comment (davidlt) [inline starlight.spec:16]: Did you forgot running make here?

Comment (KiSooLee) [inline starlight.spec:16]: make is exist below %build
make %{makeprocess}
Should I set another line for make?

Comment (davidlt) [inline starlight.spec:16]: Sorry, auto-correction changed my messages. I meant cmake.

File: starlight.spec [hunk 15]
```diff
@@ -0,0 +1,19 @@
+### RPM external starlight r193
+Requires: clhep
+
+%define branch cms/%{realversion}
+%define github_user cms-externals
+Source0:git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
```
Comment (davidlt) [inline starlight.spec]: Change `Source0` to `Source` or on `%setup` use `-b0`.

File: starlight.spec (hunk 14 above)
Comment (KiSooLee) [inline starlight.spec:16]: Actually building process is below.
cmake .
make
I just used other ".spec" file reference

**Comment by davidlt** [comment]
Do you test this before pushing the changes?

**Comment by KiSooLee** [comment]
Actually I am stuck while following the Twiki.
https://twiki.cern.ch/twiki/bin/view/CMS/SDTHowToBuildExternalTools
Can I ask you about this twiki?

**Comment by davidlt** [comment]
Simple instructions would be:

```
# Login to cmsdev0X, where X = 1, 2, 3, 4 machine 
mkdir - /build/$USER
cd /build/$USER
mkdir new_ext
cd new_ext
# you might need to change links to https ones to repositories instead
git clone -b V00-22-XX git@github.com:cms-sw/pkgtools.git PKGTOOLS
git clone -b IB/CMSSW_8_0_X/stable git@github.com:cms-sw/cmsdist.git CMSDIST
# make your changes in CMSDIST
CMS_ARCH=slc6_amd64_gcc493
screen -L time PKGTOOLS/cmsBuild -i a -a $CMS_ARCH --builders 2 -j $(getconf _NPROCESSORS_ONLN) build name_of_spec_file
```

**Comment by KiSooLee** [comment]
sorry for late reply. I was in other task.
Actually I have used only lxplus.
I have never heard about cmsdev0X before.
Could you please give me more informations about cmsdev0X?

**Comment by davidlt** [comment]
There are several virtual machines (cmsdev) managed by CMS for development purposes (CMSSW and externals), these contains all the needed packages.

**Comment by KiSooLee** [comment]
Then how can I log in to cmsdev0X?
Can all lxplus users log in to cmsdev0X?

**Comment by davidlt** [comment]
Yes, just ssh cmsdev{01,02,03,04, ..}

**Comment by davidlt** [comment]
and keep your stuff in /build/$USER folder (create one yourself), but please, cleanup once you are done. Hard drives are not made of rubber.

**Comment by KiSooLee** [comment]
Maybe I have misunderstood.
I tried several ways to login.
ssh -Y kilee@cmsdev01.cern.ch
ssh -Y kilee@cmsdev01
ssh cmsdev01
They are all failed.

**Comment by smuzaffar** [comment]
are you doing it from lxplus?

**Comment by KiSooLee** [comment]
Oh! I see.
It is internal machinary of lxplus.
Thank you!

**Comment by davidlt** [comment]
You are part of CMS group `1399(zh)`, thus a simple `ssh cmsdev01` from lxplus should work. It will authenticate you via kerberos.

**Comment by KiSooLee** [comment]
Can I ask you about source file?
Location of source file is 
https://github.com/cms-externals/starlight
And in the spec file I set
git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
The first thing makes me confusing is 'tgz'.
In the github file is uncompressed while spec file calls compressed.
And the exact name of compressed file is 'starlight-r193-src.tgz'
Then the first line of spec file should be '### RPM external starlight r193 src' I think.
And n = starlight, realversion = r193.
Then how should I naming src and use it?

**Comment by davidlt** [comment]
Source: lines are handled by build tools (in CMS it's cmsBuild), then `rpmbuild` expects to find the sources in `SOURCES` directory. So, from `rpmbuild` point of view it's `Source0: <SOURCE_DIR>/%{n}-%{realversion}.tgz`. The rest of information is just for cmsBuild to fetch sources from the correct repository, branch, tag and package them into a tarball.

`output=` tells cmsBuild how to call the tarball.

**Comment by KiSooLee** [comment]
Ok I understood it calls tarball file.
But when I try to build with the line

### RPM external starlight r193

Source:git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
I got following messages
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
Starting to process package external+starlight+r193
Checking if starlight is cached.
Package external+starlight+r193 not found in repository. Queuing for build.
Dependencies for external+starlight+r193: ['fetch-external+starlight+r193']
Fetching sources for starlight
No sources to be downloaded found for packages starlight.
(<function fetchSourcesNew at 0x1093668>, <Package name=starlight>, <scheduler.Scheduler object at 0x10c4b50>) done
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
Could you please let me know suspicious part?

**Comment by davidlt** [comment]
This clearly lacks testing.

Here are my version: https://github.com/davidlt/cmsdist/commit/c0215296c584147bf5d3a8147dc4352daec76835

It does work without CLHEP, but starlight itself needs fixing to detect CLHEP properly and compile. This is WIP version and I stop here. You can take it from make final fixes for Starlight and SPEC file should work.

**Comment by ghost** [comment]
@cmsbuild please test

File: starlight.spec [hunk 16]
```diff
@@ -0,0 +1,20 @@
+### RPM external starlight r193
+Requires: clhep
+
+%define branch cms/%{realversion}
+%define github_user cms-externals
+Source:git+http://github.com/%github_user/%{n}.git?obj=%{branch}/%{tag}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}.tgz
```
Comment (smuzaffar) [inline starlight.spec]: You are still missing space after Source: tag

**Comment by KiSooLee** [comment]
Oh, its automatically updated.
Source file is merged with new pull request.
But the spec files are now on local test.
Let me go ahead after local test of spec files.

**Comment by davidlt** [comment]
Please, use the SPEC file I provided: https://github.com/davidlt/cmsdist/commit/c0215296c584147bf5d3a8147dc4352daec76835

**Comment by KiSooLee** [comment]
Yes I'm in local test with the spec file that you have provided.
But can I ask something?
Previously tag was 36d8af7c6bb23b209e6c9bb40f3df543f85cb214.
However now source file in github is updated.
Then should I keep use 36d8af7c6bb23b209e6c9bb40f3df543f85cb214 or change to latest commit fda4f6e500e645f8907e04b5a7140368b8ed1509?

**Comment by davidlt** [comment]
Depends what was changed, you might need to modify branch & tag.

**Comment by ghost** [comment]
@KiSooLee Yes change to use the latest commit, the one starting with "fda4f6..."

**Comment by ghost** [comment]
@cmsbuild please test

**Comment by davidlt** [comment]
@KiSooLee could you use my `starlight-toolfile.spec`? It has all the issues solved.

**Comment by KiSooLee** [comment]
Yes the only difference was %build.
Now I have changed it to your starlight-toolfile.spec file.
I will commit spec files after DOXYGEN commented out starlight packge is merged.

**Comment by davidlt** [comment]
Nope, there was more differences (library name, RPM sections, whitespace added were required by RPM, etc).

**Comment by KiSooLee** [comment]
That file is different from now I'm try to local test.
I'm now copy and pasted your file and will commit this.

**Comment by KiSooLee** [comment]
starlight-toolfile.spec file is updated but starlight.spec file also need to be modified.
It will be done after the starlight package pull request have merged.
Starlight packge is modified to use DOXYGEN as option and waiting for merged.

**Comment by covarell** [comment]
@mkirsano could you please help here? It looks like this PR is going on since a long time...

**Comment by mkirsano** [comment]
OK

28/01/16 11:29, Roberto Covarelli пишет:

**Comment by davidlt** [comment]
Visual review of `starlight-toolfile.spec`, LGTM.

**Comment by KiSooLee** [comment]
Hi David
Could you please check the starlight package is reasonable also?

**Comment by davidlt** [comment]
Ir has your final changes?

**Comment by KiSooLee** [comment]
yes

**Comment by davidlt** [comment]
Will try to build and review locally this evening ir over the weekend.

**Comment by KiSooLee** [comment]
Yes, I have build starlight package locally successfully.
And it is committed already.
https://github.com/cms-externals/starlight/pull/7
Now it doesn't call DOXYGEN

**Comment by ghost** [comment]
@cmsbuild please test

**Comment by ghost** [comment]
@KiSooLee The tests ran fine and we'll discuss this PR during the next ORP meeting (02/02/2016), meanwhile, unfortunately, cmssw-tool-conf has been modified, thus making this PR un-mergeable due to conflicts. Could you please rebase it?

**Comment by smuzaffar** [comment]
#2114 is identical to this one.

**Pull request closed without merging**